mathoxide-lib/src/utils.rs

```diff
@@ -0,0 +1,77 @@
+pub struct ByRowIterator {
+    shape: Vec<usize>,
+    current: usize,
+    numel: usize,
+}
+
+impl Iterator for ByRowIterator {
+    type Item = Vec<usize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current >= self.numel {
+            None
+        } else {
+            let v = inverse_translate(self.current, &self.shape);
+            self.current += 1;
+            Some(v)
+        }
+    }
+}
+
+fn inverse_translate<Shape: AsRef<[usize]>>(idx: usize, shape: Shape) -> Vec<usize> {
+    shape.as_ref().iter().rev().scan((1, 0), |(divisor, remainder), e| {
+        let prev_div = *divisor;
+        *divisor *= e;
+        let res = ((idx - *remainder) % *divisor) / prev_div;
+        *remainder += res;
+        Some(res)
+    }).collect()
+}
```

Comment:
If I were to keep the current approach I would at least precompute the the products of the shape during the construction of the iterator e.g if `shape=[1,2,3]`, precompute `suffix_product=[6,6,3]`. Then you don't actually need to iterate here in reverse once you have the `suffix_product` precomputed